### PR TITLE
Add absolute url generation

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️

### CODE REVIEW

File: src/FastRoute.php [hunk 1]
```diff
@@ -252,6 +260,83 @@ public function generate(string $name, array $parameters = []): string
...
+     * Generates absolute URL from named route and parameters
+     *
+     * @param string $name name of the route
+     * @param array $parameters parameter-value set
+     * @param string|null $scheme host scheme
+     * @param string|null $host host for manual setup
+     * @return string URL generated
+     * @throws RouteNotFoundException in case there is no route with the name specified
+     */
+    public function generateAbsolute(string $name, array $parameters = [], string $scheme = null, string $host = null): string
```
Comment (rustamwin) [inline src/FastRoute.php]: I suggest adding the schema and host inside the parameters

Comment (yiiliveext) [reply src/FastRoute.php]: The last two parameters are optional, so in most cases we use syntax like
```php
generateAbsolute('route-name', ['id' => 42]);
```

File: tests/UrlGeneratorTest.php [hunk 2]
```diff
@@ -183,4 +193,120 @@ public function testDefaultNotUsedForRequiredParameter(): void
...
+        $request = new ServerRequest('GET', 'http://test.com/home/index');
+        $routes = [
+            Route::get('/home/index')->name('index')->host('//test.com'),
+        ];
+
+        $router = $this->createRouter($routes);
+        $router->match($request);
+        $url = $router->generateAbsolute('index');
+
+        $this->assertEquals('http://test.com/home/index', $url);
```
Comment (samdark) [inline tests/UrlGeneratorTest.php]: That doesn't look correct to me:

1. Host protocol is specified explicitly as `//`.
2. There's no override in `generateAbsolute()`.

So the result should be `//test.com/home/index`.

File: tests/UrlGeneratorTest.php [hunk 3]
```diff
@@ -183,4 +193,120 @@ public function testDefaultNotUsedForRequiredParameter(): void
...
+        $request = new ServerRequest('GET', 'http://test.com/home/index');
+        $routes = [
+            Route::get('/home/index')->name('index')->host('//mysite.com'),
+        ];
+
+        $router = $this->createRouter($routes);
+        $router->match($request);
+        $url = $router->generateAbsolute('index', [], null, '//test.com');
+
+        $this->assertEquals('http://test.com/home/index', $url);
```
Comment (samdark) [inline tests/UrlGeneratorTest.php]: Same here. I'd not expect protocol being changed to `http`.

File: src/FastRoute.php (hunk 1 above)
Comment (samdark) [reply src/FastRoute.php]: I'd not do that. There could be parameters named `host` and `scheme`.

File: tests/UrlGeneratorTest.php (hunk 3 above)
Comment (yiiliveext) [reply tests/UrlGeneratorTest.php]: > Same here. I'd not expect protocol being changed to `http`.

That's what yii2 does.
You can avoid this by setting the schema to an empty string.
```php
$router->generateAbsolute('index', [], '', '//test.com');
```

File: src/FastRoute.php (hunk 1 above)
Comment (rustamwin) [reply src/FastRoute.php]: makes sense

File: tests/UrlGeneratorTest.php (hunk 3 above)
Comment (yiiliveext) [reply tests/UrlGeneratorTest.php]: I think the removed test for overriding a scheme of the $host method param should be.

Comment (samdark) [reply tests/UrlGeneratorTest.php]: > That's what yii2 does.

Doesn't mean it's correct and expected...

Comment (samdark) [reply tests/UrlGeneratorTest.php]: > I think the removed test for overriding the scheme of the $host method param should be.
Tests should reflect every use case.

It was exact duplicate of another test.

Comment (yiiliveext) [reply tests/UrlGeneratorTest.php]: > It was exact duplicate of another test.

Which one?

Comment (yiiliveext) [reply tests/UrlGeneratorTest.php]: > > That's what yii2 does.
> 
> Doesn't mean it's correct and expected...

Ok, just imagine we fix this behavior, in what way I can override the scheme by the current scheme?

Comment (yiiliveext) [reply tests/UrlGeneratorTest.php]: I propose the next syntax 
```php
//means protocol-relative links remain protocol-relative, other will be overriden by current scheme
$router->generateAbsolute('post-index', [], '//', '//test.com');
```

Comment (samdark) [reply tests/UrlGeneratorTest.php]: Correct behavior is checking the following:

1. If `generateAbsolute()` `$scheme` parameter is specified, use that scheme
2. If there's a scheme in `generateAbsolute()` `$host`, use that scheme
3. If there's a host in route with a scheme, use that scheme
4. If there's current request available, use its scheme

Comment (samdark) [reply tests/UrlGeneratorTest.php]: > Which one?

OK, maybe it was not :) Sorry. Anyway, the points above should be ensured.

Comment (yiiliveext) [reply tests/UrlGeneratorTest.php]: > Correct behavior is checking the following:
> 
> 1. If `generateAbsolute()` `$scheme` parameter is specified, use that scheme
> 2. If there's a scheme in `generateAbsolute()` `$host`, use that scheme
> 3. If there's a host in route with a scheme, use that scheme
> 4. If there's current request available, use its scheme

Maybe it makes sense, I need to think it over.

**Comment by samdark** [comment]
👍 Thanks!

**Pull request merged**